### ISSUE TITLE
Separate spans when running Scrapy with ddtrace

### ISSUE THREAD

**Issue by AdamGold** [DataDog/dd-trace-py#2337]
### Which version of the libraries are you using?

`scrapy==2.4.1`

### How can we reproduce your problem?
When using [Scrapy](https://scrapy.org/) and trying to send traces with the `wrap` method, I get multiple traces for the same run. I tried starting a root span and appending to it, without success. This is the spider - I'm not sure it's even related to scrapy, but I am expecting these two traces to belong to the same root span:
```
class MainSpider(scrapy.Spider):
    name = "main"

    @tracer.wrap("producer.start_requests")
    def start_requests(self):
        urls = [
            "http://quotes.toscrape.com/page/1/",
        ]
        for url in urls:
            yield scrapy.Request(url=url, callback=self.parse)

    @tracer.wrap("producer.parse")
    def parse(self, response):
        return {"content": "test"}

```

running with `DATADOG_TRACE_DEBUG=true ddtrace-run scrapy crawl main`.

### What is the result that you get?
`producer.start_requests` and `producer.parse` as separate spans.

### What is the result that you expected?
`producer.start_requests` and `producer.parse` in a single trace.

**Comment by Kyle-Verhoog** [comment]
Hey again @AdamGold! I suspect this is due to some asynchronicity that scrapy is using for their callbacks. The start_requests span probably closes before self.parse is called so the spans are not connected.

To get the spans connected you can pass the span as an argument to the callback:

```python
class MainSpider(scrapy.Spider):
    name = "main"

    @tracer.wrap("producer.start_requests")
    def start_requests(self):
        urls = [
            "http://quotes.toscrape.com/page/1/",
        ]
        for url in urls:
            span = tracer.current_span()
            yield scrapy.Request(url=url, callback=self.parse, cb_kwargs={"span": span})

    def parse(self, response, span):
        with tracer.trace("producer.parse", child_of=span):
            return {"content": "test"}
```

Let me know if that works!

ref: https://docs.scrapy.org/en/latest/topics/request-response.html#scrapy.http.Request

**Comment by AdamGold** [comment]
Hey @Kyle-Verhoog, thanks for the prompt answer! I'm getting: `TypeError: trace() got an unexpected keyword argument 'child_of'`

**Comment by Kyle-Verhoog** [comment]
Ah yes my bad - I meant to use `start_span`! Look like you got what I meant though 🙂

Hmm that's really strange. When you say separate spans do you mean separate traces that show up in the UI?

**Comment by AdamGold** [comment]
@Kyle-Verhoog  Exactly, two traces without any connection between them. 

![Screen Shot 2021-04-21 at 20 33 34](https://user-images.githubusercontent.com/747847/115596501-da133d00-a2e0-11eb-8cd4-4cccf8e6be4c.png)

**Comment by Kyle-Verhoog** [comment]
Really strange! I'll have to try to repro. The snippet you posted should enforce the parenting between those spans. 🤔

**Comment by AdamGold** [comment]
@Kyle-Verhoog I got it to work with the following code, but now the built ins for Datadog are creating different traces, e.g AWS SQS:
```
def send_to_queue(content, span):
    with tracer.start_span("producer.send", child_of=span):
        client = boto3.client("sqs")
        client.send_message(
            QueueUrl=os.environ.get("QUEUE_URL"),
            MessageBody=content,
        )


class MainSpider(scrapy.Spider):
    name = "main"
    ITEM_PIPELINES = {
        "tracing.pipelines.MainPipeline": 300,
    }

    def start_requests(self):
        urls = [
            "http://quotes.toscrape.com/page/1/",
        ]
        with tracer.start_span("producer.start_requests") as span:
            for url in urls:
                yield scrapy.Request(
                    url=url, callback=self.parse, cb_kwargs={"span": span}
                )

    def parse(self, response, span):
        with tracer.start_span("producer.parse", child_of=span):
            send_to_queue("test", span)

```

Everything is on the same trace except SQS:
![Screen Shot 2021-04-23 at 11 11 49](https://user-images.githubusercontent.com/747847/115840771-b48f4c00-a424-11eb-8972-9055741d8532.png)

**Comment by Kyle-Verhoog** [comment]
Yeah so unfortunately with the current context management if you use `start_span` it creates the span in its own context (any instrumentation spans won't automatically be added). This should be addressed in #1936 but in the mean-time you'll have to manually activate the context of the spans you created. Something like this should work:

```python
def send_to_queue(content, span):
    tracer.context_provider.activate(span.context)  # add this line
    with tracer.start_span("producer.send", child_of=span):
        client = boto3.client("sqs")
        client.send_message(
            QueueUrl=os.environ.get("QUEUE_URL"),
            MessageBody=content,
        )
```

Let me know how that goes 🙂

**Comment by AdamGold** [comment]
@Kyle-Verhoog Perfect! Thanks 🙏
